assembly/src/ast/attribute.rs

```diff
@@ -0,0 +1,473 @@
+    pub fn insert_new(&mut self, attr: Attribute) -> Result<(), Attribute> {
+        if self.has(attr.name()) {
+            Err(attr)
+        } else {
+            self.insert(attr);
+            Ok(())
+        }
+    }
+
+    /// Removes the attribute named `name`, if present.
+    pub fn remove(&mut self, name: impl AsRef<str>) -> Option<Attribute> {
+        let name = name.as_ref();
+        match self.attrs.binary_search_by_key(&name, |attr| attr.name()) {
+            Ok(index) => Some(self.attrs.remove(index)),
+            Err(_) => None,
+        }
+    }
+
+    /// Clear all attributes from the set
+    #[inline]
+    pub fn clear(&mut self) {
+        self.attrs.clear();
+    }
+}
+
+impl FromIterator<Attribute> for AttributeSet {
+    #[inline]
+    fn from_iter<T: IntoIterator<Item = Attribute>>(iter: T) -> Self {
+        Self::new(iter)
+    }
+}
+
+impl Extend<Attribute> for AttributeSet {
+    fn extend<T: IntoIterator<Item = Attribute>>(&mut self, iter: T) {
+        for attr in iter {
+            self.insert(attr);
+        }
+    }
+}
+
+/// An [Attribute] represents some named metadata attached to a Miden Assembly procedure.
+///
+/// An attribute has no predefined structure per se, but syntactically there are three types:
+///
+/// * Marker attributes, i.e. just a name and no associated data. Attributes of this type are used
+///   to "mark" the item they are attached to with some unique trait or behavior implied by the
+///   name. For example, `#[inline]`.
+///
+/// * List attributes, i.e. a name and one or more comma-delimited expressions. Attributes of this
+///   type are used for cases where you want to parameterize a marker-like trait. To use a Rust
+///   example, `#[derive(Trait)]` is a list attribute, where `derive` is the marker, but we want to
+///   instruct whatever processes derives, what traits it needs to derive.
+///
+/// * Key-value attributes, i.e. a name and a value. Attributes of this type are used to attach
+///   named properties to an item. For example, `#[storage_offset = 1]`. Possible value types are:
+///   bare identifiers, decimal or hexadecimal integers, and quoted strings.
+///
+/// There are no restrictions on what attributes can exist or be used. However, there are a set of
+/// attributes that the assembler knows about, and acts on, which will be stripped during assembly.
+/// Any remaining attributes we don't explicitly handle in the assembler, will be passed along as
+/// metadata attached to the procedures in the MAST output by the assembler.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Attribute {
+    /// A named behavior, trait or action; e.g. `#[inline]`
+    Marker(Ident),
+    /// A parameterized behavior, trait or action; e.g. `#[derive(Foo)]`
+    List(MetaList),
+    /// A named property; e.g. `#[key = "value"]`, `#[key = 1]`, `#[key = 0x1]`, or `#[key = foo]`
+    KeyValue(MetaKeyValue),
+}
```

Comment:
It seems something like `#[storage(offset = 1, size = 2)]` would not be supported by this structure, right? We can of course emulate it as:
```
#[storage_offset = 1]
#[storage_size = 2]
```
But I wonder if we should try to support something like the original in the `MetaList` variant.

Reply:
While I didn't add support for that initially, it's easy to add, so I'll modify that at the same time as the syntax change in my other comment.

Reply:
You can now do either:

```
@storage(offset = 1, size = 2)
```

OR

```
@storage(size = 2)
@storage(offset = 1)
```

Key-value attributes can be declared multiple times, so long as the keys are unique in each declaration; but multiple key-value pairs can be passed in a single declaration for convenience as well. The other attribute types do _not_ permit multiple declarations.